File: bench/diff_find_bit.c
// SPDX-License-Identifier: GPL-2.0-or-later
// Tier-2.5 differential oracle: C original vs Rust translation, find_bit.
// Reference extracted from lib/find_bit.c (v7.1); kept byte-identical
// (the FIND_FIRST_BIT/FIND_NEXT_BIT/FIND_NTH_BIT macros and their C
// instantiations), host longs standing in for `unsigned long`.
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long ul;
#define BITS_PER_LONG (sizeof(ul) * 8)

#define min(a, b) ((a) < (b) ? (a) : (b))
#define BITMAP_FIRST_WORD_MASK(start) (~0UL << ((start) & (BITS_PER_LONG - 1)))
#define BITMAP_LAST_WORD_MASK(nbits) (~0UL >> (-(nbits) & (BITS_PER_LONG - 1)))

static unsigned int hweight_long(ul w)
{
	unsigned int c = 0;
	while (w) { c += w & 1; w >>= 1; }
	return c;
}

static unsigned int fns(ul word, unsigned int n)
{
	while (word && n--)
		word &= word - 1;
	return word ? __builtin_ctzl(word) : BITS_PER_LONG;
}

#define FIND_FIRST_BIT(FETCH, MUNGE, size)					\
({										\
	ul idx, val, sz = (size);						\
	for (idx = 0; idx * BITS_PER_LONG < sz; idx++) {			\
		val = (FETCH);							\
		if (val) {							\
			sz = min(idx * BITS_PER_LONG + __builtin_ctzl(MUNGE(val)), sz); \
			break;							\
		}								\
	}									\
	sz;									\
})

#define FIND_NEXT_BIT(FETCH, MUNGE, size, start)				\
({										\
	ul mask, idx, tmp, sz = (size), __start = (start);			\
	if (__start >= sz)							\
		goto out;							\
	mask = MUNGE(BITMAP_FIRST_WORD_MASK(__start));				\
	idx = __start / BITS_PER_LONG;						\
	for (tmp = (FETCH) & mask; !tmp; tmp = (FETCH)) {			\
		if ((idx + 1) * BITS_PER_LONG >= sz)				\
			goto out;						\
		idx++;								\
	}									\
	sz = min(idx * BITS_PER_LONG + __builtin_ctzl(MUNGE(tmp)), sz);	\
out:										\
	sz;									\
})

#define FIND_NTH_BIT(FETCH, size, num)						\
({										\
	ul sz = (size), nr = (num), idx, w, tmp = 0;				\
	for (idx = 0; (idx + 1) * BITS_PER_LONG <= sz; idx++) {		\
		if (idx * BITS_PER_LONG + nr >= sz)				\
			goto out;						\
		tmp = (FETCH);							\
		w = hweight_long(tmp);						\
		if (w > nr)							\
			goto found;						\
		nr -= w;							\
	}									\
	if (sz % BITS_PER_LONG)							\
		tmp = (FETCH) & BITMAP_LAST_WORD_MASK(sz);			\
found:										\
	sz = idx * BITS_PER_LONG + fns(tmp, nr);				\
out:										\
	sz;									\
})

#define NOP(x) (x)

static ul _find_first_bit(const ul *addr, ul size)
{ return FIND_FIRST_BIT(addr[idx], NOP, size); }

static ul _find_first_and_bit(const ul *addr1, const ul *addr2, ul size)
{ return FIND_FIRST_BIT(addr1[idx] & addr2[idx], NOP, size); }

static ul _find_first_andnot_bit(const ul *addr1, const ul *addr2, ul size)
{ return FIND_FIRST_BIT(addr1[idx] & ~addr2[idx], NOP, size); }

static ul _find_first_and_and_bit(const ul *addr1, const ul *addr2, const ul *addr3, ul size)
{ return FIND_FIRST_BIT(addr1[idx] & addr2[idx] & addr3[idx], NOP, size); }

static ul _find_first_zero_bit(const ul *addr, ul size)
{ return FIND_FIRST_BIT(~addr[idx], NOP, size); }

static ul _find_next_bit(const ul *addr, ul nbits, ul start)
{ return FIND_NEXT_BIT(addr[idx], NOP, nbits, start); }

static ul _find_next_and_bit(const ul *addr1, const ul *addr2, ul nbits, ul start)
{ return FIND_NEXT_BIT(addr1[idx] & addr2[idx], NOP, nbits, start); }

static ul _find_next_andnot_bit(const ul *addr1, const ul *addr2, ul nbits, ul start)
{ return FIND_NEXT_BIT(addr1[idx] & ~addr2[idx], NOP, nbits, start); }

static ul _find_next_or_bit(const ul *addr1, const ul *addr2, ul nbits, ul start)
{ return FIND_NEXT_BIT(addr1[idx] | addr2[idx], NOP, nbits, start); }

static ul _find_next_zero_bit(const ul *addr, ul nbits, ul start)
{ return FIND_NEXT_BIT(~addr[idx], NOP, nbits, start); }

static ul __find_nth_bit(const ul *addr, ul size, ul n)
{ return FIND_NTH_BIT(addr[idx], size, n); }

static ul __find_nth_and_bit(const ul *addr1, const ul *addr2, ul size, ul n)
{ return FIND_NTH_BIT(addr1[idx] & addr2[idx], size, n); }

static ul __find_nth_and_andnot_bit(const ul *addr1, const ul *addr2, const ul *addr3, ul size, ul n)
{ return FIND_NTH_BIT(addr1[idx] & addr2[idx] & ~addr3[idx], size, n); }

static ul _find_last_bit(const ul *addr, ul size)
{
	if (size) {
		ul val = BITMAP_LAST_WORD_MASK(size);
		ul idx = (size - 1) / BITS_PER_LONG;
		do {
			val &= addr[idx];
			if (val)
				return idx * BITS_PER_LONG + (BITS_PER_LONG - 1 - __builtin_clzl(val));
			val = ~0ul;
		} while (idx--);
	}
	return size;
}

// bitmap_get_value8(map, start) == bitmap_read(map, start, 8), from
// <linux/bitmap.h> (static __always_inline, specialised to nbits=8 here).
static ul bitmap_get_value8(const ul *map, ul start)
{
	size_t index = start / BITS_PER_LONG;
	ul offset = start % BITS_PER_LONG;
	ul space = BITS_PER_LONG - offset;

	if (space >= 8)
		return (map[index] >> offset) & BITMAP_LAST_WORD_MASK(8);

	ul value_low = map[index] & BITMAP_FIRST_WORD_MASK(start);
	ul value_high = map[index + 1] & BITMAP_LAST_WORD_MASK(start + 8);
	return (value_low >> offset) | (value_high << space);
}

static ul find_next_clump8(ul *clump, const ul *addr, ul size, ul offset)
{
	offset = _find_next_bit(addr, size, offset);
	if (offset == size)
		return size;
	offset = offset & ~(ul)7; // round_down(offset, 8)
	*clump = bitmap_get_value8(addr, offset);
	return offset;
}

// Explicit LCG (same constants as bench/diff_base64.c) — the shared
// project convention for reproducible cross-language input streams.
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

#define NWORDS 6
#define NBITS (NWORDS * BITS_PER_LONG)

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 424242;

	for (long i = 0; i < n; i++) {
		ul a[NWORDS], b[NWORDS], c[NWORDS];
		for (int k = 0; k < NWORDS; k++) {
			// Bias toward sparse words sometimes, dense other times,
			// to exercise both the "found in this word" and
			// "skip empty words" branches.
			uint32_t r = lcg_next();
			a[k] = (r % 4 == 0) ? 0 : ((ul)r << 32 | lcg_next());
			b[k] = (lcg_next() % 4 == 0) ? 0 : ((ul)lcg_next() << 32 | lcg_next());
			c[k] = (lcg_next() % 4 == 0) ? 0 : ((ul)lcg_next() << 32 | lcg_next());
		}
		ul size = 1 + (lcg_next() % (NBITS - 1)); // 1..NBITS-1, exercises partial last word
		ul start = lcg_next() % (size + 2); // sometimes >= size
		ul n_bit = lcg_next() % (size + 2);

		printf("first,%lu\n", _find_first_bit(a, size));
		printf("first_and,%lu\n", _find_first_and_bit(a, b, size));
		printf("first_andnot,%lu\n", _find_first_andnot_bit(a, b, size));
		printf("first_and_and,%lu\n", _find_first_and_and_bit(a, b, c, size));
		printf("first_zero,%lu\n", _find_first_zero_bit(a, size));
		printf("next,%lu\n", _find_next_bit(a, size, start));
		printf("next_and,%lu\n", _find_next_and_bit(a, b, size, start));
		printf("next_andnot,%lu\n", _find_next_andnot_bit(a, b, size, start));
		printf("next_or,%lu\n", _find_next_or_bit(a, b, size, start));
		printf("next_zero,%lu\n", _find_next_zero_bit(a, size, start));
		printf("nth,%lu\n", __find_nth_bit(a, size, n_bit));
		printf("nth_and,%lu\n", __find_nth_and_bit(a, b, size, n_bit));
		printf("nth_and_andnot,%lu\n", __find_nth_and_andnot_bit(a, b, c, size, n_bit));
		printf("last,%lu\n", _find_last_bit(a, size));

		ul clump = 0xdeadbeef;
		ul clump_off = find_next_clump8(&clump, a, size, start);
		printf("clump8,%lu,%lu\n", clump_off, clump_off == size ? 0 : clump);
	}
	return 0;
}

File: bench/diff_find_bit.rs
// SPDX-License-Identifier: GPL-2.0-or-later
//! Tier-2.5 differential oracle: Rust translation side, find_bit.
//! Faithful copy of lib/find_bit_rs.rs's algorithm (kernel-crate
//! bindings/export stripped, raw pointers replaced by slice indexing —
//! same simplification `diff_ucs2_string.rs` uses) — same protocol/LCG
//! as diff_find_bit.c.

type Ul = u64;
const BITS_PER_LONG: usize = Ul::BITS as usize;

fn first_word_mask(start: usize) -> Ul {
    Ul::MAX << (start & (BITS_PER_LONG - 1))
}

fn last_word_mask(nbits: usize) -> Ul {
    Ul::MAX >> (nbits.wrapping_neg() & (BITS_PER_LONG - 1))
}

fn fns(mut word: Ul, mut n: Ul) -> Ul {
    while word != 0 && n != 0 {
        word &= word - 1;
        n -= 1;
    }
    if word != 0 {
        word.trailing_zeros() as Ul
    } else {
        BITS_PER_LONG as Ul
    }
}

fn find_first_bit_generic(size: Ul, mut fetch: impl FnMut(usize) -> Ul) -> Ul {
    let mut idx = 0usize;
    while (idx as Ul) * (BITS_PER_LONG as Ul) < size {
        let val = fetch(idx);
        if val != 0 {
            return core::cmp::min(
                idx as Ul * BITS_PER_LONG as Ul + val.trailing_zeros() as Ul,
                size,
            );
        }
        idx += 1;
    }
    size
}

fn find_next_bit_generic(size: Ul, start: Ul, mut fetch: impl FnMut(usize) -> Ul) -> Ul {
    if start >= size {
        return size;
    }
    let mask = first_word_mask(start as usize);
    let mut idx = start as usize / BITS_PER_LONG;

    let mut tmp = fetch(idx) & mask;
    while tmp == 0 {
        if (idx + 1) as Ul * BITS_PER_LONG as Ul >= size {
            return size;
        }
        idx += 1;
        tmp = fetch(idx);
    }
    core::cmp::min(
        idx as Ul * BITS_PER_LONG as Ul + tmp.trailing_zeros() as Ul,
        size,
    )
}

fn find_nth_bit_generic(size: Ul, num: Ul, mut fetch: impl FnMut(usize) -> Ul) -> Ul {
    let mut nr = num;
    let mut idx = 0usize;
    let mut tmp: Ul = 0;

    loop {
        if (idx + 1) as Ul * BITS_PER_LONG as Ul > size {
            break;
        }
        if idx as Ul * BITS_PER_LONG as Ul + nr >= size {
            return size;
        }
        tmp = fetch(idx);
        let w = tmp.count_ones() as Ul;
        if w > nr {
            return idx as Ul * BITS_PER_LONG as Ul + fns(tmp, nr);
        }
        nr -= w;
        idx += 1;
    }
    if size % BITS_PER_LONG as Ul != 0 {
        tmp = fetch(idx) & last_word_mask(size as usize);
    }
    idx as Ul * BITS_PER_LONG as Ul + fns(tmp, nr)
}

fn find_last_bit(addr: &[Ul], size: Ul) -> Ul {
    if size == 0 {
        return size;
    }
    let mut val = last_word_mask(size as usize);
    let mut idx = (size as usize - 1) / BITS_PER_LONG;
    loop {
        val &= addr[idx];
        if val != 0 {
            return idx as Ul * BITS_PER_LONG as Ul
                + (BITS_PER_LONG as u32 - 1 - val.leading_zeros()) as Ul;
        }
        val = Ul::MAX;
        if idx == 0 {
            break;
        }
        idx -= 1;
    }
    size
}

fn bitmap_read8(map: &[Ul], start: usize) -> Ul {
    let index = start / BITS_PER_LONG;
    let offset = start % BITS_PER_LONG;
    let space = BITS_PER_LONG - offset;
    if space >= 8 {
        return (map[index] >> offset) & last_word_mask(8);
    }
    let value_low = map[index] & first_word_mask(start);
    let value_high = map[index + 1] & last_word_mask(start + 8);
    (value_low >> offset) | (value_high << space)
}

fn find_next_clump8(clump: &mut Ul, addr: &[Ul], size: Ul, offset: Ul) -> Ul {
    let offset = find_next_bit_generic(size, offset, |idx| addr[idx]);
    if offset == size {
        return size;
    }
    let offset = (offset as usize & !7) as Ul;
    *clump = bitmap_read8(addr, offset as usize);
    offset
}

// Identical LCG to diff_find_bit.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const NWORDS: usize = 6;
const NBITS: usize = NWORDS * BITS_PER_LONG;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(424242);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let mut a = [0 as Ul; NWORDS];
        let mut b = [0 as Ul; NWORDS];
        let mut c = [0 as Ul; NWORDS];
        for k in 0..NWORDS {
            let r = rng.next();
            a[k] = if r % 4 == 0 {
                0
            } else {
                (r as Ul) << 32 | rng.next() as Ul
            };
            b[k] = if rng.next() % 4 == 0 {
                0
            } else {
                (rng.next() as Ul) << 32 | rng.next() as Ul
            };
            c[k] = if rng.next() % 4 == 0 {
                0
            } else {
                (rng.next() as Ul) << 32 | rng.next() as Ul
            };
        }
        let size = 1 + (rng.next() as Ul % (NBITS as Ul - 1));
        let start = rng.next() as Ul % (size + 2);
        let n_bit = rng.next() as Ul % (size + 2);

        println!("first,{}", find_first_bit_generic(size, |idx| a[idx]));
        println!(
            "first_and,{}",
            find_first_bit_generic(size, |idx| a[idx] & b[idx])
        );
        println!(
            "first_andnot,{}",
            find_first_bit_generic(size, |idx| a[idx] & !b[idx])
        );
        println!(
            "first_and_and,{}",
            find_first_bit_generic(size, |idx| a[idx] & b[idx] & c[idx])
        );
        println!("first_zero,{}", find_first_bit_generic(size, |idx| !a[idx]));
        println!("next,{}", find_next_bit_generic(size, start, |idx| a[idx]));
        println!(
            "next_and,{}",
            find_next_bit_generic(size, start, |idx| a[idx] & b[idx])
        );
        println!(
            "next_andnot,{}",
            find_next_bit_generic(size, start, |idx| a[idx] & !b[idx])
        );
        println!(
            "next_or,{}",
            find_next_bit_generic(size, start, |idx| a[idx] | b[idx])
        );
        println!(
            "next_zero,{}",
            find_next_bit_generic(size, start, |idx| !a[idx])
        );
        println!("nth,{}", find_nth_bit_generic(size, n_bit, |idx| a[idx]));
        println!(
            "nth_and,{}",
            find_nth_bit_generic(size, n_bit, |idx| a[idx] & b[idx])
        );
        println!(
            "nth_and_andnot,{}",
            find_nth_bit_generic(size, n_bit, |idx| a[idx] & b[idx] & !c[idx])
        );
        println!("last,{}", find_last_bit(&a, size));

        let mut clump: Ul = 0xdeadbeef;
        let clump_off = find_next_clump8(&mut clump, &a, size, start);
        println!(
            "clump8,{},{}",
            clump_off,
            if clump_off == size { 0 } else { clump }
        );
    }
}
